File: source/dotnet/source/cmf.bus/cmf.bus/support/DefaultEnvelopeBus.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using Common.Logging;

using cmf.bus.core;


namespace cmf.bus.support
{
    public class DefaultEnvelopeBus : IEnvelopeBus
    {
        public event Action<Envelope, Exception> OnFailedEnvelope;


        protected SortedDictionary<int, IInboundEnvelopeProcessor> _inboundChain;
        protected SortedDictionary<int, IOutboundEnvelopeProcessor> _outboundChain;
        protected ITransportProvider _txProvider;
        protected IEnvelopeDispatcher _envDispatcher;
        protected ILog _log;

        private IDictionary<string, IList<IRegistration>> _registrar;
        private object _registrarLock = new object();


        public DefaultEnvelopeBus(
            IDictionary<int, IInboundEnvelopeProcessor> inboundProcessorChain,
            IDictionary<int, IOutboundEnvelopeProcessor> outboundProcessorChain,
            ITransportProvider transportProvider,
            IEnvelopeDispatcher envelopeDispatcher)
        {
            _inboundChain = new SortedDictionary<int, IInboundEnvelopeProcessor>(inboundProcessorChain);
            _outboundChain = new SortedDictionary<int, IOutboundEnvelopeProcessor>(outboundProcessorChain);

            _envDispatcher = envelopeDispatcher;
            _txProvider = transportProvider;
            _txProvider.OnEnvelopeReceived += new Action<Envelope>(_txProvider_OnEnvelopeReceived);

            _registrar = new Dictionary<string, IList<IRegistration>>();

            _log = LogManager.GetLogger(this.GetType());
        }


        public virtual void Send(Envelope env)
        {
            // guard clause
            if (null == env) { throw new ArgumentNullException("Cannot send a null envelope"); }
            
            // send the envelope through the outbound chain
            this.ProcessOutbound(env, _outboundChain);

            // send the envelope to the transport provider
            _txProvider.Send(env);
            
            
            _log.Info("Sent envelope: " + env.GetMessageId().ToString());
        }

        public virtual void Register(IRegistration registration)
        {
            // guard clause
            if (null == registration) { throw new ArgumentNullException("Cannot register with a null registration"); }
            
            _log.Debug("Registration received for: " + registration.Topic);
            _txProvider.Register(registration);
            _log.Info("Registered for: " + registration.Topic);
        }

        public virtual void Register(string topic, Func<Envelope, DeliveryOutcome> handler)
        {
            this.Register(new FunctionalRegistration(topic, handler));
        }


        protected virtual void ProcessOutbound(Envelope env, SortedDictionary<int, IOutboundEnvelopeProcessor> processorChain)
        {
            _log.Debug("Envelope before outbound processor chain: " + env.ToString());

            IDictionary<string, object> processorContext = new Dictionary<string, object>();

            foreach (IOutboundEnvelopeProcessor processor in processorChain.Values)
            {
                processor.ProcessOutbound(env, processorContext);
            }

            _log.Debug("Envelope after outbound processor chain: " + env.ToString());
        }

        protected virtual void ProcessInbound(Envelope env, SortedDictionary<int, IInboundEnvelopeProcessor> processorChain)
        {
            _log.Debug("Envelope before inbound processor chain: " + env.ToString());

            IDictionary<string, object> processorContext = new Dictionary<string, object>();

            foreach (IInboundEnvelopeProcessor processor in processorChain.Values)
            {
                processor.ProcessInbound(env, processorContext);
            }

            _log.Debug("Envelope after inbound processor chain: " + env.ToString());
        }

        protected virtual void _txProvider_OnEnvelopeReceived(Envelope env)
        {
            try
            {
                // get the registered handlers for topic, or null if there are none
                IEnumerable<IRegistration> registeredHandlers = this.GetRegisteredHandlers(env.GetMessageTopic());

                if (null == registeredHandlers) // there are no handlers?
                {
                    _log.Warn("An envelope arrived for which there were no registered handlers");
                }
                else // we do have handlers
                {
                    // send the envelope through the inbound processing chain
                    this.ProcessInbound(env, _inboundChain);

                    // the dispatcher encapsulates the logic of giving the envelope to handlers
                    _envDispatcher.Dispatch(env, registeredHandlers);
                }
            }
            catch (Exception ex)
            {
                _log.Error("Exception while processing an inbound message; will raise the OnFailedEnvelope event", ex);
                this.Raise_OnFailedEnvelope(env, ex);
            }
        }

        protected virtual IEnumerable<IRegistration> GetRegisteredHandlers(string topic)
        {
            IEnumerable<IRegistration> registeredHandlers = null;

            lock (_registrarLock)
            {
                if (_registrar.ContainsKey(topic))
                {
                    registeredHandlers = _registrar[topic];
                }
            }

            return registeredHandlers;
        }


        private void Raise_OnFailedEnvelope(Envelope env, Exception ex)
        {
            if (null != this.OnFailedEnvelope)
            {
                foreach (Delegate d in this.OnFailedEnvelope.GetInvocationList())
                {
                    try { d.DynamicInvoke(env, ex); }
                    catch { }
                }
            }
        }
    }
}

File: java/src/main/java/cmf/bus/berico/DefaultEnvelopeBus.java
package cmf.bus.berico;

import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import cmf.bus.Envelope;
import cmf.bus.IEnvelopeBus;
import cmf.bus.IRegistration;

public class DefaultEnvelopeBus implements IEnvelopeBus {

	protected final Logger log = LoggerFactory.getLogger(this.getClass().getCanonicalName());
	
    protected List<IInboundEnvelopeProcessor> inboundProcessors = new LinkedList<IInboundEnvelopeProcessor>();
    protected List<IOutboundEnvelopeProcessor> outboundProcessors = new LinkedList<IOutboundEnvelopeProcessor>();
    protected ITransportProvider transportProvider;

    public DefaultEnvelopeBus(ITransportProvider transportProvider) {
        this.transportProvider = transportProvider;

        initialize();
    }

    public DefaultEnvelopeBus(ITransportProvider transportProvider, List<IInboundEnvelopeProcessor> inboundProcessors,
                    List<IOutboundEnvelopeProcessor> outboundProcessors) {
        this.transportProvider = transportProvider;
        this.inboundProcessors = inboundProcessors;
        this.outboundProcessors = outboundProcessors;

        initialize();
    }

    @Override
    public void dispose() {
        transportProvider.dispose();

        for (IInboundEnvelopeProcessor p : inboundProcessors) {
            try {
                p.dispose();
            } catch (Exception ex) {
            }
        }

        for (IOutboundEnvelopeProcessor p : outboundProcessors) {
            try {
                p.dispose();
            } catch (Exception ex) {
            }
        }
    }

    void initialize() {

        // add a handler to the transport provider's envelope received event
        transportProvider.onEnvelopeReceived(
        		new InboundEnvelopeProcessorCallback(this));

        log.debug("Initialized");
    }

    boolean processInbound(Envelope envelope) {
        log.debug("Enter processInbound");

        // the inbound process can be cancelled if one of the processors returns false
        boolean noOneCancelled = true;

        Map<String, Object> context = new HashMap<String, Object>();
        for (IInboundEnvelopeProcessor inboundEnvelopeProcessor : inboundProcessors) {

        		log.trace("Processor {} is handling envelope.", inboundEnvelopeProcessor.getClass().getCanonicalName());
        	
            if (inboundEnvelopeProcessor.processInbound(envelope, context)) {
            	
                continue;
                
            } else {
            	log.info("Inbound envelope cancelled by processor of type {}", 
        				inboundEnvelopeProcessor.getClass().getCanonicalName());

                noOneCancelled = false;
                break;
            }
        }

        log.debug("Leave processInbound(noOneCancelled={})", noOneCancelled);
        return noOneCancelled;
    }

    void processOutbound(Envelope envelope) {
        log.debug("Enter processOutbound");

        Map<String, Object> context = new HashMap<String, Object>();
        for (IOutboundEnvelopeProcessor outboundEnvelopeProcessor : outboundProcessors) {
            outboundEnvelopeProcessor.processOutbound(envelope, context);
        }

        log.debug("Leave processOutbound");
    }

    @Override
    public void register(final IRegistration registration) throws Exception {
        log.debug("Enter register");
        if (registration == null) {
            throw new IllegalArgumentException("Cannot register with a null registration");
        }

        transportProvider.register(registration);
        log.debug("Leave register");
    }

    @Override
    public void send(Envelope envelope) throws Exception {
        log.debug("Enter send");
        if (envelope == null) {
            throw new IllegalArgumentException("Cannot send a null envelope");
        }

        processOutbound(envelope);
        transportProvider.send(envelope);

        log.debug("Outgoing headers: {}", new EnvelopeHelper(envelope).flatten());
        log.debug("Leave send");
    }

    public void setInboundProcessors(List<IInboundEnvelopeProcessor> inboundProcessors) {
        this.inboundProcessors = inboundProcessors;
    }

    public void setOutboundProcessorCollection(List<IOutboundEnvelopeProcessor> outboundProcessors) {
        this.outboundProcessors = outboundProcessors;
    }

    @Override
    public void unregister(IRegistration registration) throws Exception {
        log.debug("Enter unregister");

        if (registration == null) {
            throw new IllegalArgumentException("Cannot unregister with a null registration");
        }
        transportProvider.unregister(registration);

        log.debug("Leave unregister");
    }
}
